library/kani_macros/src/derive.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! This module provides Kani's derive macro for Arbitrary.
+//!
+//! ```
+//! use kani::Arbitrary;
+//!
+//! #[derive(Arbitrary)]
+//! struct S;
+//!
+//! ```
+use proc_macro2::{Ident, Span, TokenStream};
+use proc_macro_error::abort;
+use quote::{quote, quote_spanned};
+use syn::spanned::Spanned;
+use syn::{
+    parse_macro_input, parse_quote, Data, DataEnum, DeriveInput, Fields, GenericParam, Generics,
+    Index,
+};
+
+pub fn expand_derive_arbitrary(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let derive_item = parse_macro_input!(item as DeriveInput);
+    let item_name = &derive_item.ident;
+
+    // Add a bound `T: Arbitrary` to every type parameter T.
+    let generics = add_trait_bound(derive_item.generics);
+    // Generate an expression to sum up the heap size of each field.
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let body = fn_any_body(&item_name, &derive_item.data);
+    let expanded = quote! {
+        // The generated implementation.
+        impl #impl_generics kani::Arbitrary for #item_name #ty_generics #where_clause {
+            fn any() -> Self {
+                #body
+            }
+        }
+    };
+    proc_macro::TokenStream::from(expanded)
+}
+
+/// Add a bound `T: Arbitrary` to every type parameter T.
+fn add_trait_bound(mut generics: Generics) -> Generics {
+    generics.params.iter_mut().for_each(|param| {
+        if let GenericParam::Type(type_param) = param {
+            type_param.bounds.push(parse_quote!(kani::Arbitrary));
+        }
+    });
+    generics
+}
+
+/// Generate the body of the function `any()`.
+/// This will create the non-deterministic object.
+/// E.g.:
+/// ```
+/// #[derive(Arbitrary)]
+/// struct Point { x: u8, y: u8 }
+/// ```
+/// will generate the following body for `fn any()`:
+/// ```
+/// fn any() -> Self {
+///    Self { x: kani::any(), y: kani::any() }
+/// }
+/// ```
+fn fn_any_body(ident: &Ident, data: &Data) -> TokenStream {
+    match data {
+        Data::Struct(struct_data) => init_item(ident, &struct_data.fields),
+        Data::Enum(enum_data) => fn_any_enum(ident, enum_data),
+        Data::Union(_) => {
+            abort!(Span::call_site(), "Cannot derive `Arbitrary` for `{}` union", ident;
+                note = ident.span() =>
+                "The derive `Arbitrary` cannot be used for unions such as `{}`", ident
+            )
+        }
+    }
+}
+
+/// Generate an item initialization where an item can be a struct or a variant.
+/// For named fields, this will generate: `Item { field1: kani::any(), field2: kani::any(), .. }`
+/// For unnamed fields, this will generate: `Item (kani::any(), kani::any(), ..)`
+/// For unit field, generate an empty initialization.
+fn init_item(ident: &Ident, fields: &Fields) -> TokenStream {
```

Comment:
Rename to `construct_symbolic_item` or `construct_nondet_item`? The `init` gives an impression that it's getting initialized to some (concrete) value.

Reply:
Is `init_symbolic_item` OK?

Reply:
Sure.